File: android/app/src/main/kotlin/com/katala/app/bridges/ActionBridgeImpl.kt
package com.katala.app.bridges

import android.content.Context
import android.content.Intent
import android.net.Uri
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Native Android ActionBridge implementation.
 *
 * Executes external OS intents:
 * - Phone dialer via ACTION_DIAL (never ACTION_CALL)
 * - SMS pre-fill via ACTION_SENDTO
 * - Web URLs via ACTION_VIEW with strict HTTP/HTTPS validation
 */
class ActionBridgeImpl(
    private val context: Context,
    messenger: BinaryMessenger
) : MethodChannel.MethodCallHandler {

    companion object {
        const val CHANNEL_NAME = "com.katala.app/actions"
    }

    private val channel = MethodChannel(messenger, CHANNEL_NAME)

    init {
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "launchDialer" -> {
                val phoneNumber = call.argument<String>("phoneNumber")
                if (phoneNumber.isNullOrBlank()) {
                    result.error("INVALID_PHONE", "Phone number is required", null)
                    return
                }
                launchDialer(phoneNumber, result)
            }
            "launchSms" -> {
                val phoneNumber = call.argument<String>("phoneNumber")
                val message = call.argument<String>("message")
                if (phoneNumber.isNullOrBlank()) {
                    result.error("INVALID_PHONE", "Phone number is required", null)
                    return
                }
                launchSms(phoneNumber, message, result)
            }
            "launchUrl" -> {
                val url = call.argument<String>("url")
                if (url.isNullOrBlank()) {
                    result.error("INVALID_URL", "URL is required", null)
                    return
                }
                launchUrl(url, result)
            }
            "dispose" -> {
                channel.setMethodCallHandler(null)
                result.success(null)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    /**
     * Launches the system dialer pre-populated with [phoneNumber].
     * Uses ACTION_DIAL to require user confirmation before calling.
     */
    private fun launchDialer(phoneNumber: String, result: MethodChannel.Result) {
        try {
            val encodedPhone = Uri.encode(phoneNumber.trim())
            val intent = Intent(Intent.ACTION_DIAL, Uri.parse("tel:$encodedPhone")).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
            }
            context.startActivity(intent)
            result.success(true)
        } catch (e: Exception) {
            result.error("CANNOT_LAUNCH", "Failed to launch dialer: ${e.message}", null)
        }
    }

    /**
     * Opens the default SMS application with prefilled recipient and optional message.
     */
    private fun launchSms(phoneNumber: String, message: String?, result: MethodChannel.Result) {
        try {
            val encodedPhone = Uri.encode(phoneNumber.trim())
            val intent = Intent(Intent.ACTION_SENDTO, Uri.parse("smsto:$encodedPhone")).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
                if (!message.isNullOrEmpty()) {
                    putExtra("sms_body", message)
                }
            }
            context.startActivity(intent)
            result.success(true)
        } catch (e: Exception) {
            result.error("CANNOT_LAUNCH", "Failed to launch SMS: ${e.message}", null)
        }
    }

    /**
     * Opens an external web URL in the system browser.
     * Validates that the scheme is strictly HTTP or HTTPS.
     */
    private fun launchUrl(url: String, result: MethodChannel.Result) {
        try {
            val uri = Uri.parse(url.trim())
            val scheme = uri.scheme?.lowercase()

            if (scheme != "http" && scheme != "https") {
                result.error("INVALID_URL", "Only HTTP and HTTPS URLs are permitted: $url", null)
                return
            }

            val intent = Intent(Intent.ACTION_VIEW, uri).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
            }
            context.startActivity(intent)
            result.success(true)
        } catch (e: Exception) {
            result.error("CANNOT_LAUNCH", "Failed to launch URL: ${e.message}", null)
        }
    }
}

File: ios/Bridges/ActionBridgeImpl.swift
import Foundation
import Flutter
import UIKit

/**
 * Native iOS ActionBridge implementation.
 *
 * Executes OS intents:
 * - tel: URL scheme (prompts native system dialer confirmation)
 * - sms: URL scheme (opens compose sheet)
 * - http / https URLs (opens system browser, strictly rejects non-web schemes)
 */
public class ActionBridgeImpl: NSObject {

    public static let channelName = "com.katala.app/actions"

    private let channel: FlutterMethodChannel

    public init(messenger: FlutterBinaryMessenger) {
        self.channel = FlutterMethodChannel(name: ActionBridgeImpl.channelName, binaryMessenger: messenger)
        super.init()
        self.channel.setMethodCallHandler { [weak self] (call, result) in
            self?.handle(call, result: result)
        }
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "launchDialer":
            guard let args = call.arguments as? [String: Any],
                  let phoneNumber = args["phoneNumber"] as? String,
                  !phoneNumber.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                result(FlutterError(code: "INVALID_PHONE", message: "Phone number is required", details: nil))
                return
            }
            launchDialer(phoneNumber: phoneNumber, result: result)

        case "launchSms":
            guard let args = call.arguments as? [String: Any],
                  let phoneNumber = args["phoneNumber"] as? String,
                  !phoneNumber.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                result(FlutterError(code: "INVALID_PHONE", message: "Phone number is required", details: nil))
                return
            }
            let message = args["message"] as? String
            launchSms(phoneNumber: phoneNumber, message: message, result: result)

        case "launchUrl":
            guard let args = call.arguments as? [String: Any],
                  let urlString = args["url"] as? String,
                  !urlString.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                result(FlutterError(code: "INVALID_URL", message: "URL is required", details: nil))
                return
            }
            launchUrl(urlString: urlString, result: result)

        case "dispose":
            channel.setMethodCallHandler(nil)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func launchDialer(phoneNumber: String, result: @escaping FlutterResult) {
        let trimmed = phoneNumber.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let encoded = trimmed.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
              let url = URL(string: "tel:\(encoded)") else {
            result(FlutterError(code: "INVALID_PHONE", message: "Cannot construct dialer URL for \(phoneNumber)", details: nil))
            return
        }

        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:]) { success in
                if success {
                    result(true)
                } else {
                    result(FlutterError(code: "CANNOT_LAUNCH", message: "Failed to launch dialer", details: nil))
                }
            }
        }
    }

    private func launchSms(phoneNumber: String, message: String?, result: @escaping FlutterResult) {
        let trimmed = phoneNumber.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let encodedPhone = trimmed.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
            result(FlutterError(code: "INVALID_PHONE", message: "Invalid phone number: \(phoneNumber)", details: nil))
            return
        }

        var urlString = "sms:\(encodedPhone)"
        if let message = message, !message.isEmpty,
           let encodedBody = message.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
            urlString += "&body=\(encodedBody)"
        }

        guard let url = URL(string: urlString) else {
            result(FlutterError(code: "INVALID_PHONE", message: "Cannot construct SMS URL", details: nil))
            return
        }

        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:]) { success in
                if success {
                    result(true)
                } else {
                    result(FlutterError(code: "CANNOT_LAUNCH", message: "Failed to launch SMS", details: nil))
                }
            }
        }
    }

    private func launchUrl(urlString: String, result: @escaping FlutterResult) {
        let trimmed = urlString.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = URL(string: trimmed),
              let scheme = url.scheme?.lowercased() else {
            result(FlutterError(code: "INVALID_URL", message: "Invalid URL string: \(urlString)", details: nil))
            return
        }

        // Strict validation: only HTTP and HTTPS schemes are permitted
        guard scheme == "http" || scheme == "https" else {
            result(FlutterError(code: "INVALID_URL", message: "Only HTTP and HTTPS URLs are permitted. Disallowed scheme: \(scheme)", details: nil))
            return
        }

        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:]) { success in
                if success {
                    result(true)
                } else {
                    result(FlutterError(code: "CANNOT_LAUNCH", message: "Failed to open URL", details: nil))
                }
            }
        }
    }
}
